crates/starknet-types-core/src/felt/primitive_conversions.rs

```diff
@@ -0,0 +1,330 @@
+        impl From<$from> for Felt {
+            fn from(value: $from) -> Self {
+                (value as $with).into()
+            }
+        }
+    };
+}
+
+impl_from!(u8, u128);
+impl_from!(u16, u128);
+impl_from!(u32, u128);
+impl_from!(u64, u128);
+impl_from!(usize, u128);
+impl_from!(i8, i128);
+impl_from!(i16, i128);
+impl_from!(i32, i128);
+impl_from!(i64, i128);
+impl_from!(isize, i128);
+
+impl From<u128> for Felt {
+    fn from(value: u128) -> Felt {
+        Self(FieldElement::from(&UnsignedInteger::from(value)))
+    }
+}
+
+impl From<i128> for Felt {
+    fn from(value: i128) -> Felt {
+        let mut res = Self(FieldElement::from(&UnsignedInteger::from(
+            value.unsigned_abs(),
+        )));
+        if value.is_negative() {
+            res = -res;
+        }
+        res
+    }
+}
+
+// TryFrom<Felt> for primitive
+
+#[derive(Debug, Copy, Clone)]
+pub struct PrimitiveFromFeltError;
+
+#[cfg(feature = "std")]
+impl std::error::Error for PrimitiveFromFeltError {}
+
+impl core::fmt::Display for PrimitiveFromFeltError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "Failed to convert `Felt` into primitive type")
+    }
+}
+
+macro_rules! impl_try_felt_into_unsigned {
+    ($into: ty) => {
+        impl TryFrom<Felt> for $into {
+            type Error = PrimitiveFromFeltError;
+
+            fn try_from(value: Felt) -> Result<$into, Self::Error> {
+                let bytes_be = value.to_bytes_le();
+                let (bytes_to_return, bytes_to_check) =
+                    bytes_be.split_at(core::mem::size_of::<$into>());
+
+                if bytes_to_check.iter().all(|&v| v == 0) {
+                    Ok(<$into>::from_le_bytes(bytes_to_return.try_into().unwrap()))
+                } else {
+                    Err(PrimitiveFromFeltError)
+                }
+            }
+        }
+    };
+}
+
+impl_try_felt_into_unsigned!(u8);
+impl_try_felt_into_unsigned!(u16);
+impl_try_felt_into_unsigned!(u32);
+impl_try_felt_into_unsigned!(u64);
+impl_try_felt_into_unsigned!(usize);
+impl_try_felt_into_unsigned!(u128);
+
+const MINUS_TWO_BYTES_REPR: [u8; 32] = [
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 16, 0, 0, 0, 0, 0, 0, 8,
+];
+
+macro_rules! impl_try_felt_into_signed {
+    ($into: ty) => {
+        impl TryFrom<Felt> for $into {
+            type Error = PrimitiveFromFeltError;
+
+            // Let's call `s` the `size_of_type`.
+            //
+            // There is 3 ways the value we are looking for can be encoded in the Felt bytes:
+            // 1. Positive numbers:
+            //  Bytes [s..] are all set to zero, and byte [s] value must not exceed 127
+            // 2. Negative numbers:
+            //  Bytes [s..] should match `MINUS_TWO_BYTES_REPR`, and byte [s] value must not be less than 128
+            // 3. -1:
+            //  Bytes are those of Felt::MAX
```

Comment:
i don't understand that. why 128 and 127 when this macro is used for all the types

Reply:
i8::MAX 127i8    == Felt::ZERO + 127
i8::MIN -128i8 == Felt::ZERO - Felt::from(128) ==  Felt::MAX - Felt::from(127) == Felt::MAX - Felt::from(1) - Felt::from(126)

 Felt::MAX - Felt::from(1) is Felt::from(-2) and it's bytes repr is:
 ```rust
 const MINUS_TWO_BYTES_REPR: [u8; 32] = [
    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
    255, 255, 255, 255, 255, 16, 0, 0, 0, 0, 0, 0, 8,
];
```

From there all the values will just change the last bytes, reducing it's value.
So if -2i8 is encoded as 255, then
-3i8 will be 255 - 1 == 254
-12i8 will be 255 - 10 == 245
and -128 will be 255 - 126 == 129

Our computers use the `one's complement` system to represent negative numbers as bytes.
Which means that `-2i8.to_le_bytes() == [254]`, which is `255 -1`, which is the value of the last byte in `MINUS_TWO_BYTES_REPR -1`.
So to get our `-2i8` value from the bytes of `MINUS_TWO_BYTES_REPR` we just have to do `i8::from_le_bytes(MINUS_TWO_BYTES_REPR[s-1]) - 1`.
Let's see:
For `-2`: `i8::from_le_bytes([255u8; 1]) -1 == -1i8 -1 == -2`
For `i8::MIN`: `i8::from_le_bytes([129; 1]) -1 == -127i8 -1 == -128`

Now for the ` && bytes_le[size_of_type - 1] >= 0x80`.
for `i8` it's logic, if I take a value like `Felt::from(-200)`, it's byte rep will be `[57, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 16, 0, 0, 0, 0, 0, 0, 8]` and if which qualify for our ` // Negative numbers` branch. So we will get `offsetted_value == 57` and therefore parse Felt::form(-200) as `57i8`. So we need this additional check.

Now let's do the same for `i16`:
- `MIN` repr is `[1, 128, 255, ..]`,  this is parsed as `-32767`, we subtract 1 and we get `-32768`, the value we were looking for.
- `MIN -1` repr is `[0, 128, 255, ..]`, this is parsed as `-32768` and returns an error when we try the checked subtraction, which is what we want.
- `MIN -2` repr is `[255, 127, 255, ..]`, this is parsed as `32767`, we subtract 1 and get `32766` and we are fucked.

We can easily protect ourselves from this by adding the ` && bytes_le[size_of_type - 1] >= 0x80` check